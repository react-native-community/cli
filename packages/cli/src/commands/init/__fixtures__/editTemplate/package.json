{
  "name": "PlaceholderName",
  "version": "0.0.1",
  "scripts": {
    "run-ios": "react-native run-ios --scheme PlaceholderNameDevelopment"
  }
}
