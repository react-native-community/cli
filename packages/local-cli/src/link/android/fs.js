/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

const fs = require('fs-extra');

exports.readFile = file => () => fs.readFileSync(file, 'utf8');

exports.writeFile = (file, content) =>
  content
    ? fs.writeFileSync(file, content, 'utf8')
    : c => fs.writeFileSync(file, c, 'utf8');
