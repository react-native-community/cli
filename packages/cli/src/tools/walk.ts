/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import fs from 'fs';
import path from 'path';

function walk(current: string): string[] {
  if (!fs.lstatSync(current).isDirectory()) {
    return [current];
  }

  const files = fs
    .readdirSync(current)
    .map(child => walk(path.join(current, child)));
  const result: string[] = [];
  return result.concat.apply([current], files);
}

export default walk;
