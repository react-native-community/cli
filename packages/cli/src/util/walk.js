/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import fs from 'fs';
import path from 'path';

function walk(current) {
  if (!fs.lstatSync(current).isDirectory()) {
    return [current];
  }

  const files = fs
    .readdirSync(current)
    .map(child => walk(path.join(current, child)));
  return [].concat.apply([current], files);
}

module.exports = walk;
