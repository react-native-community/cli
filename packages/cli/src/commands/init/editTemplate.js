// @flow
import fs from 'fs-extra';
import path from 'path';
import walk from '../../tools/walk';
import {logger} from '@react-native-community/cli-tools';

function replaceNameInUTF8File(
  filePath: string,
  projectName: string,
  templateName: string,
) {
  logger.debug(`Replacing in ${filePath}`);

  const content = fs
    .readFileSync(filePath, 'utf8')
    .replace(new RegExp(templateName, 'g'), projectName)
    .replace(
      new RegExp(templateName.toLowerCase(), 'g'),
      projectName.toLowerCase(),
    );

  fs.writeFileSync(filePath, content, 'utf8');
}

function renameFile(filePath: string, oldName: string, newName: string) {
  const newFileName = path.join(
    path.dirname(filePath),
    path.basename(filePath).replace(new RegExp(oldName, 'g'), newName),
  );

  logger.debug(`Renaming ${filePath} -> file:${newFileName}`);

  fs.moveSync(filePath, newFileName);
}

function shouldRenameFile(filePath: string, nameToReplace: string) {
  return path.basename(filePath).includes(nameToReplace);
}

function shouldIgnoreFile(filePath: string) {
  return filePath.match(/node_modules|yarn.lock|package-lock.json/g);
}

const UNDERSCORED_DOTFILES = [
  'buckconfig',
  'eslintrc.js',
  'flowconfig',
  'gitattributes',
  'gitignore',
  'watchmanconfig',
];

function processDotfiles(filePath: string) {
  const dotfile = UNDERSCORED_DOTFILES.find(e => filePath.includes(`_${e}`));

  if (dotfile === undefined) {
    return;
  }

  renameFile(filePath, `_${dotfile}`, `.${dotfile}`);
}

export function changePlaceholderInTemplate(
  projectName: string,
  placeholderName: string,
) {
  logger.debug(`Changing ${placeholderName} for ${projectName} in template`);

  walk(process.cwd())
    .reverse()
    .forEach((filePath: string) => {
      if (shouldIgnoreFile(filePath)) {
        return;
      }
      if (!fs.statSync(filePath).isDirectory()) {
        replaceNameInUTF8File(filePath, projectName, placeholderName);
      }
      if (shouldRenameFile(filePath, placeholderName)) {
        renameFile(filePath, placeholderName, projectName);
      }
      if (shouldRenameFile(filePath, placeholderName.toLowerCase())) {
        renameFile(
          filePath,
          placeholderName.toLowerCase(),
          projectName.toLowerCase(),
        );
      }

      processDotfiles(filePath);
    });
}
