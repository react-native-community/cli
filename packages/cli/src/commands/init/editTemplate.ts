import fs from 'fs';
import path from 'path';
import {logger} from '@react-native-community/cli-tools';
import walk from '../../tools/walk';

interface PlaceholderConfig {
  projectName: string;
  placeholderName: string;
  placeholderTitle?: string;
  projectTitle?: string;
}

/**
  TODO: This is a default placeholder for title in react-native template.
  We should get rid of this once custom templates adapt `placeholderTitle` in their configurations.
*/
const DEFAULT_TITLE_PLACEHOLDER = 'Hello App Display Name';

function replaceNameInUTF8File(
  filePath: string,
  projectName: string,
  templateName: string,
) {
  logger.debug(`Replacing in ${filePath}`);
  const fileContent = fs.readFileSync(filePath, 'utf8');
  const replacedFileContent = fileContent
    .replace(new RegExp(templateName, 'g'), projectName)
    .replace(
      new RegExp(templateName.toLowerCase(), 'g'),
      projectName.toLowerCase(),
    );

  if (fileContent !== replacedFileContent) {
    fs.writeFileSync(filePath, replacedFileContent, 'utf8');
  }
}

function renameFile(filePath: string, oldName: string, newName: string) {
  const newFileName = path.join(
    path.dirname(filePath),
    path.basename(filePath).replace(new RegExp(oldName, 'g'), newName),
  );

  logger.debug(`Renaming ${filePath} -> file:${newFileName}`);

  fs.renameSync(filePath, newFileName);
}

function shouldRenameFile(filePath: string, nameToReplace: string) {
  return path.basename(filePath).includes(nameToReplace);
}

function shouldIgnoreFile(filePath: string) {
  return filePath.match(/node_modules|yarn.lock|package-lock.json/g);
}

const UNDERSCORED_DOTFILES = [
  'buckconfig',
  'eslintrc.js',
  'flowconfig',
  'gitattributes',
  'gitignore',
  'prettierrc.js',
  'watchmanconfig',
];

function processDotfiles(filePath: string) {
  const dotfile = UNDERSCORED_DOTFILES.find(e => filePath.includes(`_${e}`));

  if (dotfile === undefined) {
    return;
  }

  renameFile(filePath, `_${dotfile}`, `.${dotfile}`);
}

export function changePlaceholderInTemplate({
  projectName,
  placeholderName,
  placeholderTitle = DEFAULT_TITLE_PLACEHOLDER,
  projectTitle = projectName,
}: PlaceholderConfig) {
  logger.debug(`Changing ${placeholderName} for ${projectName} in template`);

  walk(process.cwd())
    .reverse()
    .forEach((filePath: string) => {
      if (shouldIgnoreFile(filePath)) {
        return;
      }
      if (!fs.statSync(filePath).isDirectory()) {
        replaceNameInUTF8File(filePath, projectName, placeholderName);
        replaceNameInUTF8File(filePath, projectTitle, placeholderTitle);
      }
      if (shouldRenameFile(filePath, placeholderName)) {
        renameFile(filePath, placeholderName, projectName);
      }
      if (shouldRenameFile(filePath, placeholderName.toLowerCase())) {
        renameFile(
          filePath,
          placeholderName.toLowerCase(),
          projectName.toLowerCase(),
        );
      }

      processDotfiles(filePath);
    });
}
