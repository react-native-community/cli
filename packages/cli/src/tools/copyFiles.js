/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import fs from 'fs';
import path from 'path';
import walk from './walk';

/**
 * Copy files (binary included) recursively.
 */
async function copyFiles(srcPath: string, destPath: string) {
  return Promise.all(
    walk(srcPath).map(async absoluteSrcFilePath => {
      const relativeFilePath = path.relative(srcPath, absoluteSrcFilePath);
      await copyFile(
        absoluteSrcFilePath,
        path.resolve(destPath, relativeFilePath),
      );
    }),
  );
}

/**
 * Copy a file to given destination.
 */
function copyFile(srcPath: string, destPath: string) {
  if (fs.lstatSync(srcPath).isDirectory()) {
    if (!fs.existsSync(destPath)) {
      fs.mkdirSync(destPath);
    }
    // Not recursive
    return;
  }

  return new Promise((resolve, reject) => {
    copyBinaryFile(srcPath, destPath, err => {
      if (err) {
        reject(err);
      }
      resolve(destPath);
    });
  });
}

/**
 * Same as 'cp' on Unix. Don't do any replacements.
 */
function copyBinaryFile(srcPath, destPath, cb) {
  let cbCalled = false;
  const {mode} = fs.statSync(srcPath);
  const readStream = fs.createReadStream(srcPath);
  const writeStream = fs.createWriteStream(destPath);
  readStream.on('error', err => {
    done(err);
  });
  writeStream.on('error', err => {
    done(err);
  });
  readStream.on('close', () => {
    done();
    fs.chmodSync(destPath, mode);
  });
  readStream.pipe(writeStream);
  function done(err) {
    if (!cbCalled) {
      cb(err);
      cbCalled = true;
    }
  }
}

export default copyFiles;
