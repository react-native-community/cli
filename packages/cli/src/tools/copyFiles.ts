/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs';
import path from 'path';
import walk from './walk';

type Options = {
  exclude?: Array<RegExp>;
};

/**
 * Copy files (binary included) recursively.
 */
async function copyFiles(
  srcPath: string,
  destPath: string,
  options: Options = {},
) {
  return Promise.all(
    walk(srcPath).map(async (absoluteSrcFilePath: string) => {
      const exclude = options.exclude;
      if (exclude && exclude.some(p => p.test(absoluteSrcFilePath))) {
        return;
      }
      const relativeFilePath = path.relative(srcPath, absoluteSrcFilePath);
      await copyFile(
        absoluteSrcFilePath,
        path.resolve(destPath, relativeFilePath),
      );
    }),
  );
}

/**
 * Copy a file to given destination.
 */
function copyFile(srcPath: string, destPath: string) {
  if (fs.lstatSync(srcPath).isDirectory()) {
    if (!fs.existsSync(destPath)) {
      fs.mkdirSync(destPath);
    }
    // Not recursive
    return;
  }

  return new Promise((resolve, reject) => {
    copyBinaryFile(srcPath, destPath, err => {
      if (err) {
        reject(err);
      }
      resolve(destPath);
    });
  });
}

/**
 * Same as 'cp' on Unix. Don't do any replacements.
 */
function copyBinaryFile(
  srcPath: string,
  destPath: string,
  cb: (err?: Error) => void,
) {
  let cbCalled = false;
  const {mode} = fs.statSync(srcPath);
  const readStream = fs.createReadStream(srcPath);
  const writeStream = fs.createWriteStream(destPath);
  readStream.on('error', err => {
    done(err);
  });
  writeStream.on('error', err => {
    done(err);
  });
  readStream.on('close', () => {
    done();
    fs.chmodSync(destPath, mode);
  });
  readStream.pipe(writeStream);
  function done(err?: Error) {
    if (!cbCalled) {
      cb(err);
      cbCalled = true;
    }
  }
}

export default copyFiles;
