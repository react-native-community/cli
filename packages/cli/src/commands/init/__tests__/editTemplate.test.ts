import os from 'os';
import path from 'path';
import fs from 'fs-extra';
import snapshotDiff from 'snapshot-diff';
import slash from 'slash';
import walk from '../../../tools/walk';
import copyFiles from '../../../tools/copyFiles';
import {changePlaceholderInTemplate} from '../editTemplate';

const FIXTURE_DIR = path.resolve(
  __dirname,
  '..',
  '__fixtures__',
  'editTemplate',
);
const PLACEHOLDER_NAME = 'PlaceholderName';
const PROJECT_NAME = 'ProjectName';
const PROJECT_TITLE = 'ProjectTitle';

async function createTestEnv() {
  const TEST_DIR = `rncli-should-edit-template-${Date.now()}`;
  const tmpDir = os.tmpdir();
  const testPath = path.resolve(tmpDir, TEST_DIR);

  fs.mkdirSync(testPath);
  await copyFiles(FIXTURE_DIR, testPath);

  return testPath;
}

let testPath;

beforeEach(async () => {
  testPath = await createTestEnv();
});

afterEach(() => {
  fs.removeSync(testPath);
});

test('should edit template', () => {
  jest.spyOn(process, 'cwd').mockImplementation(() => testPath);

  changePlaceholderInTemplate({
    projectName: PROJECT_NAME,
    placeholderName: PLACEHOLDER_NAME,
  });

  const transformedTree = walk(testPath).map(e => e.replace(testPath, ''));
  const fixtureTree = walk(FIXTURE_DIR).map(e => e.replace(FIXTURE_DIR, ''));

  const oldJavaFile = fs.readFileSync(
    path.resolve(
      FIXTURE_DIR,
      'android',
      'com',
      PLACEHOLDER_NAME.toLowerCase(),
      'Main.java',
    ),
    'utf8',
  );
  const newJavaFile = fs.readFileSync(
    path.resolve(
      testPath,
      'android',
      'com',
      PROJECT_NAME.toLowerCase(),
      'Main.java',
    ),
    'utf8',
  );

  const oldCFile = fs.readFileSync(
    path.resolve(FIXTURE_DIR, 'ios', PLACEHOLDER_NAME, 'AppDelegate.m'),
    'utf8',
  );
  const newCFile = fs.readFileSync(
    path.resolve(testPath, 'ios', PROJECT_NAME, 'AppDelegate.m'),
    'utf8',
  );

  expect(snapshotDiff(oldCFile, newCFile, {contextLines: 1})).toMatchSnapshot();
  expect(
    snapshotDiff(oldJavaFile, newJavaFile, {contextLines: 1}),
  ).toMatchSnapshot();
  expect(
    snapshotDiff(fixtureTree.map(slash), transformedTree.map(slash), {
      contextLines: 1,
    }),
  ).toMatchSnapshot();
});

test('should edit template with custom title', () => {
  jest.spyOn(process, 'cwd').mockImplementation(() => testPath);

  changePlaceholderInTemplate({
    projectName: PROJECT_NAME,
    placeholderName: PLACEHOLDER_NAME,
    projectTitle: PROJECT_TITLE,
  });

  const oldJavaFile = fs.readFileSync(
    path.resolve(
      FIXTURE_DIR,
      'android',
      'com',
      PLACEHOLDER_NAME.toLowerCase(),
      'Main.java',
    ),
    'utf8',
  );
  const newJavaFile = fs.readFileSync(
    path.resolve(
      testPath,
      'android',
      'com',
      PROJECT_NAME.toLowerCase(),
      'Main.java',
    ),
    'utf8',
  );

  expect(
    snapshotDiff(oldJavaFile, newJavaFile, {contextLines: 1}),
  ).toMatchSnapshot();
});
