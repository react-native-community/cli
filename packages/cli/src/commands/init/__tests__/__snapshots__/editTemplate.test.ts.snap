// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should edit template 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -12,3 +12,3 @@
    RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
-                                             moduleName:@\\"PlaceholderName\\"
+                                             moduleName:@\\"ProjectName\\"
                                              initialProperties:nil];"
`;

exports[`should edit template 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,2 +1,2 @@
- com.placeholdername;
+ com.projectname;

@@ -4,4 +4,4 @@
    public static void run() {
-     String name = \\"PlaceholderName\\";
-     String title = \\"Hello App Display Name\\";
+     String name = \\"ProjectName\\";
+     String title = \\"ProjectName\\";
    }"
`;

exports[`should edit template 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -4,12 +4,12 @@
    \\"/android/com\\",
-   \\"/android/com/placeholdername\\",
-   \\"/android/com/placeholdername/Main.java\\",
-   \\"/android/com/placeholdername/PlaceholderName.java\\",
+   \\"/android/com/projectname\\",
+   \\"/android/com/projectname/Main.java\\",
+   \\"/android/com/projectname/ProjectName.java\\",
    \\"/ios\\",
-   \\"/ios/PlaceholderName\\",
-   \\"/ios/PlaceholderName/AppDelegate.m\\",
-   \\"/ios/PlaceholderName-tvOS\\",
-   \\"/ios/PlaceholderName-tvOS/.gitkeep\\",
-   \\"/ios/PlaceholderNameTests\\",
-   \\"/ios/PlaceholderNameTests/.gitkeep\\",
+   \\"/ios/ProjectName\\",
+   \\"/ios/ProjectName/AppDelegate.m\\",
+   \\"/ios/ProjectName-tvOS\\",
+   \\"/ios/ProjectName-tvOS/.gitkeep\\",
+   \\"/ios/ProjectNameTests\\",
+   \\"/ios/ProjectNameTests/.gitkeep\\",
    \\"/node_modules\\","
`;

exports[`should edit template with custom title 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,2 +1,2 @@
- com.placeholdername;
+ com.projectname;

@@ -4,4 +4,4 @@
    public static void run() {
-     String name = \\"PlaceholderName\\";
-     String title = \\"Hello App Display Name\\";
+     String name = \\"ProjectName\\";
+     String title = \\"ProjectTitle\\";
    }"
`;
